### PR TITLE
Build docker image on Circle CI

### PR DESCRIPTION
**Pull request by Squareys**
As per https://github.com/emscripten-core/emscripten/issues/11078 and https://github.com/emscripten-core/emsdk/issues/372.

@trzecieu @sbc100 
Very very rough draft for building the docker image on Circle CI.

**TODO**
 - [x] Build docker image
 - [x] Properly specify the tag(s) for docker hub
 - [x] Trigger build only on ~~`emscripten-releases-tags.txt` changes.~~ tags
    - [x] Double check that this is _really_ just triggering on tags.
 - [x] What's the process of getting the docker hub deploy key? They should be specified through secret variables in Circle CI and probably could be done as a final step by one of the project maintainers.
 - ~~Build various variations of the image? (i.e. I remember emscripten-slim and a ubuntu based version in trzeci's original images.)~~

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -90,6 +90,15 @@ jobs:
             source emsdk_env.sh
             python scripts/test.py
 
+  build-docker-image:
+    machine: true
+    steps:
+      - checkout
+      - run: docker build -t emscripten-core/emsdk:$CIRCLE_BRANCH docker/
+      - run: |
+          echo "$DOCKER_PASS" | docker login --username $DOCKER_USER --password-stdin
```
Comment (sbc100) [inline .circleci/config.yml:99]: Is this the recommended way to authenticate the push ?  I wonder if we can pass login credentials directly to the push command somehow?

Comment (Squareys) [reply .circleci/config.yml:99]: This is from the CircleCI docs, yes. It is definitely possible to pass the password directly:

```
docker login -u "$DOCKER_USER" -p "$DOCKER_PASS" ...
```

Comment (sbc100) [reply .circleci/config.yml:99]: Pattern from the CircleCI docs, seems fine

Comment (trzecieu) [reply .circleci/config.yml:99]: Looks odd, but also I've found that the recommended - more secure way for login.

File: .circleci/config.yml [hunk 2]
```diff
@@ -90,6 +90,15 @@ jobs:
             source emsdk_env.sh
             python scripts/test.py
 
+  build-docker-image:
+    machine: true
+    steps:
+      - checkout
+      - run: docker build -t emscripten-core/emsdk:$CIRCLE_BRANCH docker/
```
Comment (trzecieu) [inline .circleci/config.yml:97]: We do have docker.io organization called [emscripten/emsdk](https://hub.docker.com/repository/docker/emscripten/emsdk/general). 
I don't know who manages this organisation, looks like @haraldreingruber has access to another project [emscripten/emscripten-ci](https://hub.docker.com/repository/docker/emscripten/emscripten-ci/general) - maybe he knows more :)

Comment (sbc100) [reply .circleci/config.yml:97]: I seem to remember deciding on `emscripten/emsdk` for this image, right?  Or am I remembering wrong?

Comment (trzecieu) [reply .circleci/config.yml:97]: Most likely you're right. And looking at the name of this repo it makes sense. 
In this case there should be a following action on https://hub.docker.com/orgs/emscripten

**Comment by Squareys** [comment]
>  Trigger build only on emscripten-releases-tags.txt changes.

@sbc100 It looks like whenever this file changes, emsdk is tagged. Could we instead make this job trigger on tags instead? That would be very straight forward to achieve.

File: .circleci/config.yml (hunk 1 above)
Comment (Squareys) [reply .circleci/config.yml:101]: > Error: Cannot perform an interactive login from a non TTY device

Looks like `--password-stdin` is not accepted here anyways? 🤔

**Comment by Squareys** [comment]
From my side this is pretty much ready, what's left is:

 - Ensure this is running only on tags (I set up a filter which should make the job ignored for all branches, but included for all tags). The `$CIRCLE_TAG` variable is only set for tags, hence it fails for this PR.
 - Set `DOCKER_PASS` and `DOCKER_USER` secret variables in the Circle CI settings after merge.

So, this is ready to be merged from my side.

File: .circleci/config.yml [hunk 3]
```diff
@@ -90,9 +90,19 @@ jobs:
...
+      - run: cd docker && make version=${CIRCLE_TAG}-upstream
+      - run: |
+          docker login -u "$DOCKER_USER" -p "$DOCKER_PASS"
+          docker push emscripten/emsdk:${CIRCLE_TAG}-upstream
+
 workflows:
+  version: 2
   flake8:
-    jobs:
+    jobs:ignore: /.*/
```
Comment (sbc100) [inline .circleci/config.yml:105]: What does this do?

File: .circleci/config.yml [hunk 4]
```diff
@@ -90,9 +90,19 @@ jobs:
             source emsdk_env.sh
             python scripts/test.py
 
+  build-docker-image:
+    machine: true
```
Comment (sbc100) [inline .circleci/config.yml:94]: Why this?

File: .circleci/config.yml [hunk 5]
```diff
@@ -90,9 +90,19 @@ jobs:
...
+    steps:
+      - checkout
+      - run: cd docker && make version=${CIRCLE_TAG}-upstream
+      - run: |
```
Comment (sbc100) [inline .circleci/config.yml:98]: Does this run at the top level, or within the `docker` subdirectory?   (i.e. does the `cd` in the previous run apply to this run too?

File: .circleci/config.yml (hunk 4 above)
Comment (Squareys) [reply .circleci/config.yml:94]: This uses a docker machine executor which is required to build docker images.
https://circleci.com/docs/2.0/examples-intro/#linux-with-machine

File: .circleci/config.yml (hunk 3 above)
Comment (Squareys) [reply .circleci/config.yml:105]: Sorry, that was a pretty sloppy mistake on my behalf

File: .circleci/config.yml (hunk 5 above)
Comment (Squareys) [reply .circleci/config.yml:98]: I am pretty sure that docker push and login do not require the Docker file but use the tag to reference the build docker image stored in the local registry. Right, @trzecieu?

Comment (trzecieu) [reply .circleci/config.yml:98]: That's right, push and login can be done from any place.

File: .circleci/config.yml (hunk 2 above)
Comment (haraldreingruber) [reply .circleci/config.yml]: The org (https://hub.docker.com/orgs/emscripten) was created by @kripken, right?
It's odd, but I don't see any other members...

**Comment by Squareys** [comment]
@sbc100 I fixed or commented on the issues you found. Tell me if there is anything else I can do.

**Comment by sbc100** [comment]
lgtm.

What are the next steps?  It looks like the CI is currently failing... and I guess we need to figure out how to inject the credentials?

**Comment by Squareys** [comment]
Injecting the credentials should be done after merge by someone who has access to the Circle CI backend and the docker hub namespace (@kripken? See https://github.com/emscripten-core/emsdk/pull/513#discussion_r435738602).

The CI is failing because it is not running for a tag, hence `CIRCLE_TAG` variable is missing aswell. Once the PR is merged, it will only run on tags.

**Comment by sbc100** [comment]
> The CI is failing because it is not running for a tag, hence `CIRCLE_TAG` variable is missing aswell. Once the PR is merged, it will only run on tags.

It seems a little odd to me that this would run on this PR at all.. if its only supposed to run on tags.. why is it running here?

**Pull request merged**